### PR TITLE
issue/289/documentation

### PR DESCRIPTION
**Pull request by combet**
On GitHub page: check and/or update
- [x] README.md
- [x] OVERVIEW.md
- [x] CONTRIBUTING.md
- [x] Move detailed installation instructions out of CONTRIBUTING into a specific INSTALL page.
- [x] Add mention to the possibility to start install from the DESC conda env for DESC members, at NERSC and CC-IN2P3. 

On the online documentation page: (need to run `./update_docs` and then open `docs/_build/html/index.html` to see how it renders)
- [x] Update installation instructions
- [x] Add an overview section, similar to the one on the github repo. 
- [x] ~~Add an overview section there as well ?~~ Actually, the list of examples available online are sufficient to give an idea of what the CLMM does

### CODE REVIEW

**Comment by hsinfan1996** [comment]
FoFCatalogMatching can be installed with pip.

**Comment by hsinfan1996** [comment]
ccl is also pip installable with `pip install pyccl`

**Comment by combet** [comment]
> ccl is also pip installable with `pip install pyccl`

I doesn't always install with pip, but the conda install seems more robust. Added this to the installation documentation.

File: INSTALL.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+Note, you may choose to install some or all of the ccl, numcosmo, and/or cluster_toolkit packages.  You need at least one.  If you install cluster_toolkit and others, then you need to install cluster_toolkit *last*.   If you have already installed cluster_toolkit before the other packages, simply run, `pip uninstall cluster_toolkit` then re-install cluster_toolkit.
+
+Now, you can install CLMM and its dependencies as
+
+```bash
+    pip install numpy scipy astropy matplotlib
+    pip install pytest sphinx sphinx_rtd_theme
+    pip install jupyter  # need to have jupyter notebook tied to this environment, you can then see the environment in jupyter.nersc.gov
+    git clone https://github.com/LSSTDESC/CLMM.git  # For those with edit access to CLMM, see below on how to fork the repo otherwise
```
Comment (hsinfan1996) [inline INSTALL.md:44]: without edit access?

**Comment by hsinfan1996** [comment]
`g_t =& -\left( g_1\cos\left(2\phi\right)-g_2\sin\left(2\phi\right)\right)\\` is this correct? The sign of g2 seems to be wrong

**Comment by combet** [comment]
> `g_t =& -\left( g_1\cos\left(2\phi\right)-g_2\sin\left(2\phi\right)\right)\\` is this correct? The sign of g2 seems to be wrong

Nice catch! That's in the docstrings of `compute_tangential_and_cross_components`, right? I fixed that one at least.

File: INSTALL.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+If you are a DESC member you may also add to your CLMM environment the GCR and GCRCatalog packages to access the DC2 datasets at NERSC. To run the DC2 example notebooks provided in CLMM, the following need to be installed in your CLMM environment at NERSC. Once in your CLMM environment (`source activate clmmenv`), run
+
+```bash
+    pip install pandas
+    pip install pyarrow
+    pip install healpy
+    pip install h5py
+    pip install GCR
+    pip install https://github.com/LSSTDESC/gcr-catalogs/archive/master.zip
+    pip install https://github.com/yymao/FoFCatalogMatching/archive/master.zip
```
Comment (hsinfan1996) [inline INSTALL.md:80]: or simply `pip install FoFCatalogMatching`. I tried on nersc and it is working fine.

File: INSTALL.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+## Making a local copy of CLMM <a name="making_a_local_copy_of_clmm"></a>
+
+As a newcomer, you likely will not have edit access to the main CLMM repository.
+Without edit privileges, you won't be able to create or push changes to branches in the base repository. You can get around this by creating a [fork](https://help.github.com/articles/fork-a-repo/), a linked copy of the CLMM repository under your Github username. You can then push code changes to your fork which can later be merged with the base repository.
+
+To create a fork, navigate to the [CLMM home page](https://github.com/LSSTDESC/CLMM) and click 'Fork' in the upper right hand corner. The fork has been created under your username on Github's remote server and can now be cloned to your local repository with
+
+```bash
+    git clone git@github.com:YOUR-USERNAME/CLMM.git
```
Comment (hsinfan1996) [inline INSTALL.md:118]: Do I need to do `cd CLMM` before the next step?

File: docs/source/installation.rst [hunk 4]
```diff
@@ -14,34 +14,31 @@ To run the tests you can do::
   
 Requirements
 ============
+CLMM requires Python version 3.6 or later.  CLMM has the following dependencies:
 
-CLMM requires Python version 3.6 or later.  To run the code, there are the following dependencies:
+- `numpy <http://www.numpy.org/>`_: 1.17 or later
```
Comment (hsinfan1996) [inline docs/source/installation.rst:19]: NumPy

File: README.md [hunk 5]
```diff
@@ -1,46 +1,44 @@
...
-3. [Contributing](#contributing)
+4. [Contributing](#contributing)
+5. [Contact](#contact)
 
 ## Requirements <a name="requirements"></a>
 
-CLMM requires Python version 3.6 or later.  To run the code, there are the following dependencies:
+CLMM requires Python version 3.6 or later.  CLMM has the following dependencies:
 
 - [numpy](http://www.numpy.org/) (1.17 or later)
```
Comment (hsinfan1996) [inline README.md:22]: Referring as NumPy if we want to be really accurate.

File: README.md [hunk 6]
```diff
@@ -1,46 +1,44 @@
...
-
 - [scipy](http://www.numpy.org/) (1.3 or later)
```
Comment (hsinfan1996) [inline README.md:23]: ...SciPy...

File: README.md [hunk 7]
```diff
@@ -1,46 +1,44 @@
...
-
 - [astropy](https://www.astropy.org/) (3.x or later for units and cosmology dependence)
```
Comment (hsinfan1996) [inline README.md:24]: ...Astropy...

File: README.md [hunk 8]
```diff
@@ -1,46 +1,44 @@
...
-
 - [matplotlib](https://matplotlib.org/) (for plotting and going through tutorials)
```
Comment (hsinfan1996) [inline README.md:25]: ...Matplotlib...

File: README.md [hunk 9]
```diff
@@ -1,46 +1,44 @@
...
+- [CCL](https://ccl.readthedocs.io/en/v2.0.0/)
+- [NumCosmo](https://numcosmo.github.io/) 
+
+(See the [INSTALL documentation](INSTALL.md) for more detailed installation instructions.)
 
 For developers, you will also need to install:
 
 - [pytest](https://docs.pytest.org/en/latest/) (3.x or later for testing)
-
 - [sphinx](https://www.sphinx-doc.org/en/master/usage/installation.html) (for documentation)
```
Comment (hsinfan1996) [inline README.md:42]: ...Sphinx...

Comment (hsinfan1996) [inline README.md:42]: ...Sphinx...

File: README.md (hunk 8 above)
Comment (hsinfan1996) [inline README.md:25]: ...Matplotlib...

File: INSTALL.md (hunk 2 above)
Comment (combet) [reply INSTALL.md]: Changed.

File: INSTALL.md (hunk 3 above)
Comment (combet) [reply INSTALL.md:118]: Yes, fixed.

File: README.md (hunk 5 above)
Comment (combet) [reply README.md]: Done.

File: README.md (hunk 6 above)
Comment (combet) [reply README.md]: Done.

File: README.md (hunk 7 above)
Comment (combet) [reply README.md]: Done.

File: README.md (hunk 9 above)
Comment (combet) [reply README.md]: Done.

File: README.md (hunk 8 above)
Comment (combet) [reply README.md]: Done.

**Review by hsinfan1996 — APPROVED** [review]

**Comment by hsinfan1996** [comment]
Not sure if it is just me or it is written somewhere in the documentation, I need cluster-toolkits installed in order for sphinx to build the documentation.

**Comment by m-aguena** [comment]
> Not sure if it is just me or it is written somewhere in the documentation, I need cluster-toolkits installed in order for sphinx to build the documentation files.

That is because sphinx needs execute all files to make the documentation. I don't know if we can/want to put a failsafe on that. @combet @pennalima thoughts?

**Comment by hsinfan1996** [comment]
> > Not sure if it is just me or it is written somewhere in the documentation, I need cluster-toolkits installed in order for sphinx to build the documentation files.
> 
> That is because sphinx needs execute all files to make the documentation. I don't know if we can/want to put a failsafe on that. @combet @pennalima thoughts?
I thought any of the backends was equivalent. Maybe it is not the case for building documentation.

**Comment by combet** [comment]
> > > Not sure if it is just me or it is written somewhere in the documentation, I need cluster-toolkits installed in order for sphinx to build the documentation files.
> > 
> > 
> > That is because sphinx needs execute all files to make the documentation. I don't know if we can/want to put a failsafe on that. @combet @pennalima thoughts?
> 
> I thought any of the backends was equivalent. Maybe it is not the case for building documentation.

I'll try to check this out, but I thought as well that it would be equivalent...

**Comment by m-aguena** [comment]
@combet @hsinfan1996 this happens because we want the documentation to be built for all of the code, so it has to "go" into every `.py` file and you need an environment with everything working.

**Comment by hsinfan1996** [comment]
> @combet @hsinfan1996 this happens because we want the documentation to be built for all of the code, so it has to "go" into every `.py` file and you need an environment with everything working.

Yes, this part I understand. I am wondering why I need cluster-toolkit particularly. Is it because cluster-toolkit is the default backend? I tried building the documentation with only CCL installed and it failed but succeeded with cluster-toolkit installed.

**Comment by m-aguena** [comment]
That is strange, the documentation should fail if any backend is missing,

**Comment by hsinfan1996** [comment]
Just did some test, only need cluster-toolkit installed

**Comment by combet** [comment]
Following up on that, in #371, @pennalima switch the default backend to `ccl` instead of `cluster-toolkit`, and the documentation does not build on GA (see, [here](https://github.com/LSSTDESC/CLMM/pull/371/checks?check_run_id=1788516915)) where cluster-toolkit is installed. 

I tried locally, setting up the banckend env variable to  `ccl` and reproduced the error. A similar error arise if `numcosmo` is defined as the backend. Commenting out the `assert` lines causing the error, the documentation build OK if cluster-toolkit is also installed.

**Comment by vitenti** [comment]
@m-aguena @combet The solution is to use mock imports, for cluster_toolkit this was already done in docs/conf.py using the code:
```
    # For these modules, do a mock import
    MOCK_MODULES = ['cluster_toolkit']
    sys.modules.update((mod_name, Mock()) for mod_name in MOCK_MODULES)
```
I don't have time to do it right now, one needs just to add ccl and related modules to MOCK_MODULES and it should work.

**Comment by m-aguena** [comment]
@vitenti I will try it.

**Comment by m-aguena** [comment]
@vitenti what is the purpose of:
```
MOCK_MODULES = ['gi', 'gi.repository', 'gi.repository.NumCosmoMath', 'gi.repository.NumCosmo', 'pyccl']
for mod_name in MOCK_MODULES:
    sys.modules[mod_name] = MagicMock()
```
in `docs/conf.py` ? I found that removing these lines solves the problems building the documentation.

**Comment by m-aguena** [comment]
@vitenti I found a good workaround modifying only the `docs/conf.py` file. Check PR #371

**Review by m-aguena — APPROVED** [review]

**Pull request merged**